### ISSUE TITLE
Operations: Please allow setting nodeSelector, tolerations, and affinity on all Pods

### ISSUE THREAD

**Issue by AceHack** [knative/operator#5]
Most companies with complex enough kubernetes deployments that have multiple different node groups require some sort of topology selection so pods will run on the correct group of nodes.  Please allow setting nodeSelector, tolerations, and affinity on all Pods to accomplish this.  This includes adding it to any CRDs that create pods so these values are settable.

https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/

See below for example:
```
apiVersion: v1
kind: Pod
metadata:
  name: sample
spec:
  nodeSelector:
    disktype: sad
  tolerations:
  - key: "example-key"
    operator: "Exists"
    effect: "NoSchedule"
  affinity:
    nodeAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - weight: 1
        preference:
          matchExpressions:
          - key: another-node-label-key
            operator: In
            values:
            - another-node-label-value
```

Related Feature Request: https://github.com/knative/eventing/issues/2883

**Comment by houshengbo** [comment]
This is what is implemented in other operator:

NodeSelector: https://github.com/helm/charts/blob/master/stable/prometheus-operator/crds/crd-prometheus.yaml#L3223

Affinity: https://github.com/helm/charts/blob/master/stable/prometheus-operator/crds/crd-prometheus.yaml#L119

Tolerations: https://github.com/helm/charts/blob/master/stable/prometheus-operator/crds/crd-prometheus.yaml#L4523

**Comment by AceHack** [comment]
We found PodNodeSelector and PodTolerationRestriction admission controllers.
https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#podnodeselector
https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#podtolerationrestriction
https://github.com/kubernetes/kubernetes/issues/57424

But most hosted K8s control plane providers do not support them because their annotations are alpha.
https://docs.aws.amazon.com/eks/latest/userguide/platform-versions.html

We also found this open-source project that allows for setting node selector in namespace annotations using a webhook.
https://github.com/liangrog/admission-webhook-server

We will likely try to contribute the other tolerations, and affinity back to this project assuming the maintainer is open to it.

**Comment by AceHack** [comment]
Is it possible to follow a pattern like in serving for setting node selectors on the serving pods such as activator, autoscale, controller, etc...

https://github.com/knative/serving/issues/1816

**Comment by AceHack** [comment]
Bump

**Comment by adamkgray** [comment]
@AceHack you mean setting tolerations and node selectors for the knative serving components (autoscaler, activator etc.) via the KnativeServing CRD from the Operator, correct?

**Comment by nak3** [comment]
I think we should start with a global config like current `spec.high-availability`.
I am thinking that we can create `spec.global` then add some configurations under it like:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  global:
    nodeSelector:
      environment: dev
```

Users might want to customize for a specific deployment but we should start with this global one.
If no objections or other opinions, I will start creating the pull request.

**Comment by markusthoemmes** [comment]
@nak3 shouldn't this be part of the existing `deployments` section maybe?

**Comment by nak3** [comment]
@markusthoemmes Ah yes, `deployments` section would be good. 
So it will be needed to configure each deployment like: 
```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  deployments:
  - name: webhook
    nodeSelector:
      environment: dev
  - name: controller
    nodeSelector:
      environment: dev
```

**Comment by markusthoemmes** [comment]
I guess we could add a `global` or `all` convenience section to `deployments`, which'd still allow to set them to all at once.

**Comment by nak3** [comment]
Umm... The `deployments` is a `[]array` type so I think we cannot add the `global` or `all` section there?
https://github.com/knative/operator/blob/1bbdd9004e64757c27f6be0c612ee3973db8cb1d/pkg/apis/operator/v1alpha1/common.go#L134
https://github.com/knative/operator/blob/1bbdd9004e64757c27f6be0c612ee3973db8cb1d/pkg/apis/operator/v1alpha1/common.go#L220

**Comment by isaac88** [comment]
Hello @nak3 @evankanderson guys,

Do you have any ETA to allow us put tolerations in the Knative operator? 

For us it's blocking because we have the same issue of @AceHack 

Thanks

**Comment by nak3** [comment]
@isaac88 Thank you for the information. I think we are alright to add it. I sent the PR https://github.com/knative/operator/pull/747 so it will be available on the next release (v0.26) if it was merged.

**Comment by nak3** [comment]
AAll three `nodeSelector`, `tolerations` and `affinity` are now implemented. (`tolerations` and `affinity` will be available after 0.26+).
